src/experiments/udpsocket/api.js

```diff
@@ -0,0 +1,90 @@
+"use strict";
+/* exported udpsocket */
+/* global Components, ChromeUtils, ExtensionCommon, ExtensionAPI */
+
+const Cc = Components.classes;
+const Ci = Components.interfaces;
+const { Services } = ChromeUtils.import("resource://gre/modules/Services.jsm");
+const { EventManager} = ExtensionCommon;
+
+const socket_ipv4_a         = Cc["@mozilla.org/network/udp-socket;1"].createInstance(Ci.nsIUDPSocket);
```

Comment:
The reason why I added seven UDPSocket instances is because, without decoding the responses I receive, I need some way of associating each response with the RRTYPE that was queried. Furthermore, if we fail to receive a response (e.g. due to middlebox interference, which is what we're studying), then I need some way to associate the lack of a response with RRTYPE that was queried.

If I continue to not send queries in parallel, then I suppose I could store the RRTYPE that was queried in a global variable in `dns-test.js`. However, that doesn't feel clean either.

Regarding the 5 second timeout, we want to mimic the behavior of a stub resolver that sends a UDP query but fails to receive a response. By default, resolvconf waits 5 seconds after sending a UDP query before re-trying. It's not immediately clear to me how we can both a) wait up to 5 seconds before re-trying, and b) return immediately if we receive a response.

Reply:
I think that all of these questions have been answered by my previous comments from this round of review, but if I missed one, feel free to ping me again.

Reply:
I simplified the udpsocket API in 5b36d0f. I still need to add timeout management; will do that in a bit.

Reply:
In my code snippet with the promise constructor, I showed an example of how you can handle timeouts: https://github.com/mozilla-extensions/dnssec-interference/pull/5#discussion_r469494933

In background.js, `setTimeout` is available by default.

In this experiment script, you need to import it from `resource://gre/modules/Timer.jsm` (search on [Searchfox](https://searchfox.org/mozilla-central/search?q=resource%3A%2F%2Fgre%2Fmodules%2FTimer.jsm&path=&case=false&regexp=false) for examples).

Reply:
Addressed by 7707b94